api/src/main/java/org/openmrs/module/registrationcore/api/mpi/openempi/OpenEmpiImplementation.java

```diff
@@ -0,0 +1,78 @@
+package org.openmrs.module.registrationcore.api.mpi.openempi;
+
+import org.openmrs.Patient;
+import org.openmrs.module.registrationcore.api.mpi.common.*;
+import org.openmrs.module.registrationcore.api.search.PatientAndMatchQuality;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+
+import java.util.List;
+import java.util.Map;
+
+public class OpenEmpiImplementation implements MpiProvider {
+
+    @Autowired
+    @Qualifier("registrationcore.mpiPatientImporter")
+    private MpiPatientImporter patientImporter;
+
+    @Autowired
+    @Qualifier("registrationcore.mpiPatientExporter")
+    private MpiPatientExporter patientExporter;
+
+    @Autowired
+    @Qualifier("registrationcore.mpiPatientUpdater")
+    private MpiPatientUpdater patientUpdater;
+
+    @Autowired
+    @Qualifier("registrationcore.mpiPatientSearcher")
+    private MpiSimilarPatientsSearcher searchAlgorithm;
+
+    @Autowired
+    @Qualifier("registrationcore.mpiAuthenticator")
+    private MpiAuthenticator authenticator;
+
+    @Autowired
+    @Qualifier("registrationcore.mpiProperties")
+    private MpiProperties mpiProperties;
+
+    @Override
+    public Patient importMpiPatient(String patientId) {
+        validateAuthentication();
+        return patientImporter.importMpiPatient(patientId);
+    }
+
+    @Override
+    public OpenEmpiPatientQuery exportPatient(Patient patient) {
+        validateAuthentication();
+        return patientExporter.exportPatient(patient);
+    }
+
+    @Override
+    public void updatePatient(Patient patient) {
+        validateAuthentication();
+        patientUpdater.updatePatient(patient);
+    }
+
+    @Override
+    public List<PatientAndMatchQuality> findProbablySimilarPatients(Patient patient, Map<String, Object> otherDataPoints,
+                                                                    Double cutoff, Integer maxResults) {
+        validateAuthentication();
+        if (mpiProperties.isProbablyMatchingEnabled()) {
+            return searchAlgorithm.findProbablySimilarPatients(patient, otherDataPoints, cutoff, maxResults);
```

Comment:
seeing this, it feels wrong to me that a single "searchAlgorithm" has _both_ the similar and exact queries within it. I would expect to have one "searchAlgorithm" for each